src/openslide-vendor-philips-isyntax.c

```diff
@@ -0,0 +1,397 @@
+        level->base.w = libisyntax_level_get_width_in_tiles(level->isyntax_level) * level->base.tile_w;
+        level->base.h = libisyntax_level_get_height_in_tiles(level->isyntax_level) * level->base.tile_h;
+        osr->levels[i] = (struct _openslide_level*)level;
+        level->grid = _openslide_grid_create_simple(
+                osr,
+                libisyntax_level_get_width_in_tiles(level->isyntax_level),
+                libisyntax_level_get_height_in_tiles(level->isyntax_level),
+                level->base.tile_w,
+                level->base.tile_h,
+                philips_isyntax_read_tile);
+
+        // TODO(avirodov): log levels? Right now too noisy.
+        // LOG_VAR("%d", i);
+        // LOG_VAR("%d", libisyntax_level_get_scale(level->isyntax_level));
+        // LOG_VAR("%d", libisyntax_level_get_width_in_tiles(level->isyntax_level));
+        // LOG_VAR("%d", libisyntax_level_get_height_in_tiles(level->isyntax_level));
+        // LOG_VAR("%f", libisyntax_level_get_downsample_factor(level->isyntax_level));
+        // TODO(avirodov): getters in API, if needed.
+        // LOG_VAR("%f", levels[i].um_per_pixel_x);
+        // LOG_VAR("%f", levels[i].um_per_pixel_y);
+        // LOG_VAR("%f", levels[i].x_tile_side_in_um);
+        // LOG_VAR("%f", levels[i].y_tile_side_in_um);
+        // LOG_VAR("%lld", levels[i].tile_count);
+        // LOG_VAR("%f", levels[i].origin_offset_in_pixels);
+        // LOG_VAR("%f", levels[i].origin_offset.x);
+        // LOG_VAR("%f", levels[i].origin_offset.y);
+        // LOG_VAR("%d", (int)levels[i].is_fully_loaded);
+    }
+    osr->ops = &philips_isyntax_ops;
+    return true;
+}
+
+static bool philips_isyntax_paint_region(
+        openslide_t *osr G_GNUC_UNUSED, cairo_t *cr,
+        int64_t x, int64_t y,
+        struct _openslide_level *osr_level,
+        int32_t w, int32_t h,
+        GError **err) {
+    philips_isyntax_level* level = (philips_isyntax_level*)osr_level;
+
+    // LOG("x=%ld y=%ld level=%d w=%d h=%d", x, y, level->level_idx, w, h);
+    // Note: round is necessary to avoid producing resampled (and thus blurry) images on higher levels.
+    double origin_offset_in_pixels = libisyntax_level_get_origin_offset_in_pixels(level->isyntax_level);
+    return _openslide_grid_paint_region(level->grid, cr, NULL,
+                                        round((x - origin_offset_in_pixels) / level->base.downsample),
+                                        round((y - origin_offset_in_pixels) / level->base.downsample),
+                                        osr_level, w, h,
+                                        err);
+}
+
+static void philips_isyntax_destroy(openslide_t *osr) {
+    philips_isyntax_t *data = osr->data;
+
+    for (int i = 0; i < osr->level_count; ++i) {
+        philips_isyntax_level* level = (philips_isyntax_level*)osr->levels[i];
+        _openslide_grid_destroy(level->grid);
+        free(level);
```

Comment:
Level freeing should go in a separate function that can be used with `g_autoptr()`.

Reply:
Done.